src/lmdb.ml

```diff
@@ -167,21 +170,88 @@ module Env = struct
 
 end
 
+module Txn :
+sig
+  type -'a t = mdb_txn constraint 'a = [< `Read | `Write ]
+
+  val go :
+    ?parent:([< `Read | `Write ] as 'a) t ->
+    ?rw:'a cap ->
+    Env.t ->
+    ('a t -> 'b) -> 'b option
+
+  val abort : 'a t -> 'b
+
+  val env : 'a t -> Env.t
+
+  (* not exported: *)
+  val maybe_txn :
+    ?rw:'a cap ->
+    ?txn:'a t ->
+    Env.t ->
+    ('a t -> 'b) -> 'b
+end
+=
+struct
+  type -'a t = mdb_txn constraint 'a = [< `Read | `Write ]
+
+  exception Abort of mdb_txn
+
+  let env txn = mdb_txn_env txn
+
+  let abort txn = raise (Abort txn)
+
+  let go :
+    ?parent: 'a t ->
+    ?rw:'a cap ->
+    Env.t ->
+    ('a t -> 'b) -> 'b option
+    =
+    fun ?parent ?rw env f ->
+    let ptr_txn = alloc mdb_txn in
+    let txn_flag = match rw with
+      | Some `Read -> Env.Flags.read_only
+      | Some `Write -> Env.Flags.none
+      | None -> Env.Flags.none
+    in
+    he@@mdb_txn_begin env parent txn_flag ptr_txn ;
+    let txn = !@ptr_txn in
+    try
+      let x = f txn in
+      he@@mdb_txn_commit txn ; Some x
+    with
+      | Abort t' when t' == txn || parent = None ->
+        mdb_txn_abort txn ; None
+      | exn -> mdb_txn_abort txn ; raise exn
+
+  (* Used internally for trivial functions, not exported. *)
+  let maybe_txn :
+    ?rw:'a cap ->
+    ?txn:'a t ->
+    Env.t ->
+    ('a t -> 'b) -> 'b
+    =
+    fun ?rw ?txn e ->
+    match txn with
+    | Some txn ->
+      let e' = env txn in
+      if ptr_compare e' e <> 0
+      then invalid_arg
+          "Lmdb: database and transaction are from different environments";
+      (|>) txn
```

Comment:
Ugh, please don't play that kind of games, we are not in Haskell. Write all the function arguments.

Reply:
```suggestion
"Lmdb: Db.t and Txn.t are from different environments";
```
I'm not sure what you have in mind.
We could also just drop the check and let ``mdb.c`` report the error.